internal/controllers/machineset/machineset_controller_status.go

```diff
@@ -0,0 +1,278 @@
+	if machines == nil {
+		return
+	}
+
+	var readyReplicas, availableReplicas int32
+	for _, machine := range machines {
+		if meta.IsStatusConditionTrue(machine.GetV1Beta2Conditions(), clusterv1.MachineReadyV1Beta2Condition) {
+			readyReplicas++
+		}
+		if meta.IsStatusConditionTrue(machine.GetV1Beta2Conditions(), clusterv1.MachineAvailableV1Beta2Condition) {
+			availableReplicas++
+		}
+		if !machine.GetDeletionTimestamp().IsZero() && time.Since(machine.GetDeletionTimestamp().Time) >= time.Minute*30 {
+			s.staleDeletingMachines = append(s.staleDeletingMachines, machine.GetName())
+		}
+	}
+
+	if ms.Status.V1Beta2 == nil {
+		ms.Status.V1Beta2 = &clusterv1.MachineSetV1Beta2Status{}
+	}
+
+	ms.Status.Replicas = int32(len(machines))
+	ms.Status.V1Beta2.ReadyReplicas = ptr.To(readyReplicas)
+	ms.Status.V1Beta2.AvailableReplicas = ptr.To(availableReplicas)
+}
+
+func setScalingUpCondition(ms *clusterv1.MachineSet, machines []*clusterv1.Machine, bootstrapObjectNotFound, infrastructureObjectNotFound bool) {
+	// If we got unexpected errors in listing the machines (this should happen rarely), surface them
+	if machines == nil {
+		v1beta2conditions.Set(ms, metav1.Condition{
+			Type:    clusterv1.MachineSetScalingUpV1Beta2Condition,
+			Status:  metav1.ConditionUnknown,
+			Reason:  clusterv1.MachineSetScalingUpErrorV1Beta2Reason,
+			Message: "Please check controller logs for errors",
+		})
+		return
+	}
+
+	// Surface if .spec.replicas is not yet set (this should happen rarely).
+	// This could e.g. be the case when using autoscaling with MachineSets.
+	if ms.Spec.Replicas == nil {
+		v1beta2conditions.Set(ms, metav1.Condition{
+			Type:   clusterv1.MachineSetScalingUpV1Beta2Condition,
+			Status: metav1.ConditionUnknown,
+			Reason: clusterv1.MachineSetWaitingForReplicasToBeSetV1Beta2Reason,
+		})
+		return
+	}
+
+	desiredReplicas := *ms.Spec.Replicas
+	gotReplicas := ms.Status.Replicas
+
+	referencesMessage := calculateMissingReferencesMessage(bootstrapObjectNotFound, infrastructureObjectNotFound)
+
+	// Not scaling up or in deletion.
+	if gotReplicas >= desiredReplicas || !ms.DeletionTimestamp.IsZero() {
+		var message string
+		if referencesMessage != "" {
+			message = fmt.Sprintf("ScalingUp would fail because %s", referencesMessage)
+		}
+		v1beta2conditions.Set(ms, metav1.Condition{
+			Type:    clusterv1.MachineSetScalingUpV1Beta2Condition,
+			Status:  metav1.ConditionFalse,
+			Reason:  clusterv1.MachineSetNotScalingUpV1Beta2Reason,
+			Message: message,
+		})
+		return
+	}
+
+	message := fmt.Sprintf("ScalingUp from %d to %d replicas", gotReplicas, desiredReplicas)
+	if referencesMessage != "" {
+		message = fmt.Sprintf("%s is blocked because %s", message, referencesMessage)
+	}
+	// Scaling up.
+	v1beta2conditions.Set(ms, metav1.Condition{
+		Type:    clusterv1.MachineSetScalingUpV1Beta2Condition,
+		Status:  metav1.ConditionTrue,
+		Reason:  clusterv1.MachineSetScalingUpV1Beta2Reason,
+		Message: message,
```

Comment:
Not sure if we want it or not, but also in this case we might want to surface machine provisioning (without a node) by more than 30 minutes.

Reply:
Hm but this doesn't matter for scale up right? I think it's confusing to include this information here if it doesn't matter for scale up to complete

Reply:
agree, the above should surface in the MachinesReady condition?!